source/log_channel.c

```diff
@@ -0,0 +1,260 @@
+
+    aws_thread_join(&impl->background_thread);
+
+    aws_thread_clean_up(&impl->background_thread);
+    aws_condition_variable_clean_up(&impl->pending_line_signal);
+    aws_array_list_clean_up(&impl->pending_log_lines);
+    aws_mutex_clean_up(&impl->sync);
+    aws_mem_release(channel->allocator, impl);
+
+    return AWS_OP_SUCCESS;
+}
+
+static struct aws_log_channel_vtable s_background_channel_vtable = {
+    .send_fn = s_background_channel_send_fn,
+    .cleanup_fn = s_background_channel_cleanup_fn
+};
+
+static bool s_background_wait_fn(void *context) {
+    struct aws_log_background_channel *impl = (struct aws_log_background_channel *)context;
+
+    /*
+     * Condition variable predicates are checked under mutex protection
+     */
+    return impl->finished || aws_array_list_length(&impl->pending_log_lines) > 0;
+}
+
+static void s_background_thread_writer_fn(void *thread_data) {
+    (void) thread_data;
+
+    struct aws_log_channel *channel = (struct aws_log_channel *)thread_data;
+    struct aws_log_background_channel *impl = (struct aws_log_background_channel *)channel->impl;
+
+    struct aws_array_list log_lines;
+    if (aws_array_list_init_dynamic(&log_lines, channel->allocator, 10, sizeof(struct aws_string *))) {
+        /*
+         * This feels like it should be a fatal error, as it's invisible to the user and it will result
+         * in an endless backup of log lines, assuming they are able to be created.
+         */
+        return;
+    }
+
+    while(true) {
+        aws_mutex_lock(&impl->sync);
+        aws_condition_variable_wait_pred(&impl->pending_line_signal, &impl->sync, s_background_wait_fn, impl);
+
+        size_t line_count = aws_array_list_length(&impl->pending_log_lines);
+        bool finished = impl->finished;
+
+        if (line_count == 0) {
+            aws_mutex_unlock(&impl->sync);
+            if (finished) {
+                break;
+            }
+            continue;
+        }
+
+        aws_array_list_swap_contents(&impl->pending_log_lines, &log_lines);
+        aws_mutex_unlock(&impl->sync);
+
+        /*
+         * Consider copying these into a page-sized stack buffer (string) and then making the write calls
+         * against it rather than the individual strings.  Might be a savings when > 1 lines (cut down on
+         * write calls).
+         */
+        for (size_t i = 0; i < line_count; ++i) {
+            struct aws_string *log_line = NULL;
+            if (aws_array_list_get_at(&log_lines, &log_line, i)) {
+                continue;
+            }
+
+            (*channel->writer->vtable->write_fn)(channel->writer, log_line);
+
+            /*
+             * send is considered a transfer of ownership.  write is not a transfer of ownership.
+             * So it's always the channel's responsibility to clean up all log lines that enter
+             * it as soon as they are no longer needed.
+             */
+            aws_string_destroy(log_line);
```

Comment:
These should definitely be pooled and pre-allocated to a decent capacity. Otherwise the churn will be magnificent, though at least the runtime of it is eaten by the logger thread.

Reply:
In addition to the above comment, one other thing is that at the very least, we need a simple baseline implementation against which we can do performance comparisons for all the improvements we end up making (like the previous comment, for example).  

I may change the naming convention from "default" to "standard" and then add an options struct where we throw in switches for all of the optimizations.